### PR TITLE
feat(react): port React chart wrappers to TypeScript

### PR DESCRIPTION
**Pull request by metonym**
### Updates

- Port React chart wrappers to TypeScript (`.jsx` --> `.tsx`)
- Install TypeScript and React type definitions as development dependencies
- Update the build script (`react/build.sh`) to use TypeScript to build library in CJS format and emit Type declarations
- Update TypeScript entry in package.json to be `"./index.d.ts"`
- Update `data` type in core Chart interfaces to reflect the new tabular data format (`Record<string, any>[]`)
- Update React storybook configuration to use transpiled TS when running in development mode or building for production

### Notes

- Some fields in chart `options` will require the consumer to use the ENUM type (e.g. [ScaleTypes](https://github.com/metonym/carbon-charts/blob/e91808cc7432838826484bed763475034d586287/packages/react/carbon__charts-react-tests.tsx#L34)) from the core `@carbon/charts` package

### Review checklist (for reviewers only)
- [ ] Demos all features
- [ ] Documented/annotated
- [ ] Matches UI/UX specs
- [ ] Meets the [code style guide](https://github.com/carbon-design-system/carbon-charts/wiki/Code-style-guide)
- [ ] Accessible
- [ ] Mobile first (responsive)
- [ ] RTL support (bidirectional text)
- [ ] Performant (limited bloat)

### CODE REVIEW

**Review by cal-smith — COMMENTED** [review]
It may be easier to maintain if we just switch the react codebase over to typescript ... That way all the type declarations can be co-located with the code, and we won't have to worry about syncing them. I'm not sure what we're using for builds right now, but if it's babel we should be able to enable Typescript parsing there and just use tsc to generate the `.d.ts`

**Comment by theiliad** [comment]
> It may be easier to maintain if we just switch the react codebase over to typescript ... That way all the type declarations can be co-located with the code, and we won't have to worry about syncing them. I'm not sure what we're using for builds right now, but if it's babel we should be able to enable Typescript parsing there and just use tsc to generate the `.d.ts`

That makes sense, any comments @metonym?

**Comment by metonym** [comment]
@theiliad I concur. Porting to TypeScript would be easier to maintain and less error prone.

File: packages/react/src/line-chart.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import { LineChart as LC } from "@carbon/charts";
+import {
+  ChartData,
+  LineChartOptions,
+} from "@carbon/charts/interfaces";
+import BaseChart, { TabularData } from "./base-chart";
+
+type LineChartData =
+  | TabularData
+  | ChartData
```
Comment (theiliad) [inline packages/react/src/line-chart.tsx:11]: we actually need to remove this type or update it to match the tabular format. Not sure why it's still in our types

Comment (theiliad) [reply packages/react/src/line-chart.tsx:11]: all mentions of `datasets` & `labels` in the charting data should also disappear from the react types or tests

Comment (theiliad) [reply packages/react/src/line-chart.tsx:11]: https://carbon-design-system.github.io/carbon-charts/?path=/story/tutorials--tabular-data-format

Comment (metonym) [reply packages/react/src/line-chart.tsx]: Added a `ChartTabularData` type to the core model interfaces, which is pretty loose: `Record<string, any>[]`  

Haven't yet removed mentions of `datasets/labels` etc..

File: packages/react/package.json [hunk 2]
```diff
@@ -76,6 +79,7 @@
 		"sass": "1.25.0",
 		"sass-loader": "8.0.0",
 		"style-loader": "0.19.0",
+		"typescript": "^4.1.2",
```
Comment (theiliad) [inline packages/react/package.json:82]:
```suggestion
		"typescript": "4.1.2",
```

File: packages/react/package.json [hunk 3]
```diff
@@ -56,6 +57,8 @@
 		"@storybook/addon-knobs": "5.3.12",
 		"@storybook/addon-options": "5.3.12",
 		"@storybook/react": "5.3.12",
+		"@types/react": "^17.0.0",
+		"@types/react-dom": "^17.0.0",
```
Comment (theiliad) [inline packages/react/package.json:61]:
```suggestion
		"@types/react-dom": "17.0.0",
```

File: packages/react/package.json [hunk 4]
```diff
@@ -56,6 +57,8 @@
 		"@storybook/addon-knobs": "5.3.12",
 		"@storybook/addon-options": "5.3.12",
 		"@storybook/react": "5.3.12",
+		"@types/react": "^17.0.0",
```
Comment (theiliad) [inline packages/react/package.json:60]:
```suggestion
		"@types/react": "17.0.0",
```

File: packages/react/carbon__charts-react-tests.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+              "Dataset 3": "#725699",
+              "Dataset 4": "#ccc",
+            },
+          },
+          height: "400px",
+        }}
+      />
+    );
+  }
+}
```
Comment (theiliad) [inline packages/react/carbon__charts-react-tests.tsx:55]:
```suggestion
}

```

File: packages/react/carbon__charts-react-tests.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import { LineChart } from "./dist";
+
+class App extends React.Component {
+  ref: React.RefObject<LineChart> = React.createRef();
+
+  state = {
+    data: {
+      labels: [''],
+      datasets: []
+    }
```
Comment (theiliad) [inline packages/react/carbon__charts-react-tests.tsx:12]: this is the older data format that we deprecated many versions ago https://carbon-design-system.github.io/carbon-charts/?path=/story/tutorials--tabular-data-format

File: packages/react/src/base-chart.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+  shouldComponentUpdate(nextProps: Props<Options, Data>) {
+    return (
+      this.props.data !== nextProps.data ||
+      this.props.options !== nextProps.options
+    );
+  }
+
+  componentDidUpdate() {
+    // @ts-ignore
```
Comment (theiliad) [inline packages/react/src/base-chart.tsx:44]: curios as to why these are needed? do we need to fix something in the `setData` method? add some typing maybe?

Comment (theiliad) [reply packages/react/src/base-chart.tsx:44]: not necessarily requesting a change here, more so looking for the problem

**Review by theiliad — CHANGES_REQUESTED** [review]

File: packages/react/build.sh [hunk 8]
```diff
@@ -9,12 +9,12 @@ echo "linking local packages (non-lerna)"
...
+echo "compiling ts to js"
+tsc
 
 echo "bundling..."
 rollup -c
 
 echo "copying metadata"
 cp *.md dist/
-cp package.json dist/
+cp package.json dist/
```
Comment (theiliad) [inline packages/react/build.sh:20]:
```suggestion
cp package.json dist/

```

**Review by theiliad — APPROVED** [review]

**Review by theiliad — CHANGES_REQUESTED** [review]
meant to request the change

File: packages/react/carbon__charts-react-tests.tsx (hunk 6 above)
Comment (theiliad) [reply packages/react/carbon__charts-react-tests.tsx:12]: are you able to update this to the tabular format?

Comment (metonym) [reply packages/react/carbon__charts-react-tests.tsx]: Updated in https://github.com/carbon-design-system/carbon-charts/pull/867/commits/e9a53f4336a7af42a6b9caf555185df5624fc4de#diff-02ba3e7463b36e80fe42c20ff1e8d59c9575c7dbbd322bb6ce24630a2e240272

File: packages/react/src/base-chart.tsx (hunk 7 above)
Comment (metonym) [reply packages/react/src/base-chart.tsx]: Fixed in https://github.com/carbon-design-system/carbon-charts/pull/867/commits/e9a53f4336a7af42a6b9caf555185df5624fc4de#diff-827cd8b812b357745e517c845fa7d5ca9f9d0cb61fdb0e50d285f809ff9d631b

Comment (theiliad) [reply packages/react/src/base-chart.tsx]: Nice 🎉

**Review by theiliad — CHANGES_REQUESTED** [review]
I'm good to merge, except that the react storybook is fully breaking. Are you able to fix @metonym? 🙂

**Review by theiliad — APPROVED** [review]
💯

**Review by cal-smith — APPROVED** [review]
Sweet 🎉 Nice job getting this switched over @metonym

**Pull request merged**